File: code-examples/java-rs/crud/geospatial.java
import com.mongodb.ConnectionString;
import com.mongodb.reactivestreams.client.MongoClients;
import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoCollection;
import com.mongodb.reactivestreams.client.MongoDatabase;
import org.bson.Document;
import com.mongodb.client.model.geojson.Point;
import com.mongodb.client.model.geojson.Polygon;
import com.mongodb.client.model.geojson.Position;
import com.mongodb.client.model.Filters;
import com.mongodb.Block;

import java.util.Arrays;

public class NearAndWithinPolygon {
    private static final ConnectionString URI = new ConnectionString("mongodb+srv://<user>:<password>@<cluster-url>?writeConcern=majority");
    
    public static void main(final String[] args) {

        final MongoClient client = MongoClients.create(URI);
        final MongoDatabase database = client.getDatabase("sample_mflix");
        final MongoCollection<Document> theaters = database.getCollection("theaters");
        
        // Find theaters within a certain proximity
        Block<Document> printBlock = document -> System.out.println(document.toJson());
        Point refPoint = new Point(new Position(-73.9667, 40.78));
        theaters.find(Filters.near("location.geo", refPoint, 10000.0, null)).forEach(printBlock).subscribe();

        // Find theaters within a specific geographic range
        Position position1 = new Position(-72, 40);
        Position position2 = new Position(-74, 41);
        Position position3 = new Position(-72, 39);
        Position position4 = new Position(-72, 40);
        Polygon polygon = new Polygon(Arrays.asList(position1, position2, position3, position4));
        theaters.find(Filters.geoWithin("location.geo", polygon)).forEach(printBlock).subscribe();

        // Close the client after the query completes
        client.close();
    }
}

File: code-examples/scala/crud/geospatial.scala
import org.mongodb.scala._
import org.mongodb.scala.bson._
import org.mongodb.scala.model.Filters._
import org.mongodb.scala.model.geojson._

import scala.concurrent.Await
import scala.concurrent.duration._

object GeospatialQueries {

  val mongoClient: MongoClient = MongoClient("mongodb+srv://<user>:<password>@<cluster-url>?writeConcern=majority")
  val database: MongoDatabase = mongoClient.getDatabase("sample_mflix")
  val theaters: MongoCollection[Document] = database.getCollection("theaters")

  // Find theaters within a certain proximity
  def proximity(): Unit = {
    val point = new Point(new Position(-73.9667, 40.78))
    val query = near("location.geo", point, 10000.0, 0.0)

    val observable = theaters.find(query)

    observable.subscribe((doc: Document) => println(doc.toJson()))

    if (observable.toFuture().isCompleted && observable.toFuture().value.get.isSuccess && observable.toFuture().value.get.get.isEmpty) {
      println("No documents found!")
    }
  }

  // Find theaters within a specific geographic range
  def range(): Unit = {
    val polygon = new Polygon(
      new Position(-72, 40), // Polygon coordinates defining the range
      new Position(-74, 41),
      new Position(-72, 39),
      new Position(-72, 40),
    )
    val query = geoWithin("location.geo", polygon)

    val observable = theaters.find(query)

    observable.subscribe((doc: Document) => println(doc.toJson()))

    if (observable.toFuture().isCompleted && observable.toFuture().value.get.isSuccess && observable.toFuture().value.get.get.isEmpty) {
      println("No documents found!")
    }
  }

  def run(): Unit = {
    try {
      proximity()
      range()
    } finally {
      mongoClient.close()
    }
  }

  def main(args: Array[String]): Unit = {
    run()
  }
}
